luasocket5.3 : initial add

cwluxul opened the pull request:
Creating a luasocket package that is compatible with, and built against,
lua version 5.3.  The base of this package started from the luasocket
package and I made the following modifications:

Makefile - Updated the upstream version.  Set the LUAV environment
variable in the build to 5.3.  Set an include path in the lua build so
that that lua5.3 headers could be found.  Also using the make install
target as everything is put in the correct place automatically.

0001-Add-interface-support.patch - ported this patch which included
replacing one deprecated call.  Validated that the interface
functionality added by the patch is working and functional.  The
following lua script line was functional: print(socket.iface.indextoname(2))

0301-Fix-mpc85xx-build.patch - Just fixed the offsets and fuzz in the
patch

040-remove-fpic-and-warnings.patch - ported this patch.  I did make one
change.  The patch name is called remove fpic and warnings, but the
patch was also removing optimization.  I ported the removal of fpic and
warnings, however I left optimization in.

Built and tested on BCM5301X.

Signed-off-by: Colby Whitney <colby.whitney@luxul.com>

Maintainer: me / @\<github-user> (find it by checking history of the package Makefile)
Compile tested: (put here arch, model, OpenWrt version)
Run tested: (put here arch, model, OpenWrt version, tests done)

Description:

neheb commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,58 @@
+
+PKG_NAME:=luasocket
+PKG_SOURCE_VERSION:=733af884f1aa18ff469bf3c4d18810e815853211
+PKG_VERSION:=3.0-rc1-20190421
```
Remove PKG_VERSION and add PKG_SOURCE_DATE. Use https://github.com/openwrt/packages/blob/master/utils/usbmuxd/Makefile as a guide

cwluxul replied:
I've switched to PKG_SOURCE_DATE

neheb commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,58 @@
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
+PKG_SOURCE_URL:=https://github.com/diegonehab/luasocket.git
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
```
Remove

cwluxul replied:
I've removed PKG_SOURCE_SUBDIR

neheb commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,58 @@
+PKG_VERSION:=3.0-rc1-20190421
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
```
Remove

cwluxul replied:
I've removed PKG_SOURCE

neheb commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,59 @@
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
+PKG_MIRROR_HASH:=915b06219cb79802fe547c1b1896bd7c5465b717549d6a0f38e2c0b4f3e7ae3e
+PKG_SOURCE_URL:=https://github.com/diegonehab/luasocket.git
```
Remove .git

cwluxul replied:
I've removed the .git

BKPepe commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,59 @@
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=LuaSocket 5.3
+  URL:=http://luasocket.luaforge.net/
```
404 Not found

cwluxul replied:
This is a copy from the normal luasocket makefile.  Should I update to: http://w3.impa.br/~diego/software/luasocket/
?

neheb replied:
Yes

cwluxul replied:
This should now be updated.

neheb commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,59 @@
+  CATEGORY:=Languages
+  TITLE:=LuaSocket 5.3
+  URL:=http://luasocket.luaforge.net/
+  MAINTAINER:=W. Michael Petullo <mike@flyn.org>
```
Similar to the linked package, remove and add PKG_MAINTAINER above as well as license information.

cwluxul replied:
@neheb - I don't understand the request "remove and add PKG_MAINTAINER".  I'm confused when you say to remove it and add it at the same time.  I'll add the license.

neheb replied:
Check https://github.com/openwrt/packages/blob/master/utils/usbmuxd/Makefile

There's PKG_MAINTAINER and PKG_LICENSE at the top, not down here.

cwluxul replied:
I've moved the PKG_MAINTAINER and added PKG_LICENSE

neheb replied:
Place PKG_MAINTAINER above PKG_LICENSE

karlp replied:
While PKG_MAINTAINER works, the documented way is still MAINTAINER.  Please don't demand people change them "just because"  See https://openwrt.org/docs/guide-developer/packages

cwluxul replied:
@neheb - Per the "conversation" above, I was informed PKG_MAINTAINER can be blank.  I've done that, thus I think putting it above PKG_LICENSE is now no longer valid.

BKPepe commented on lang/luasocket5.3/Makefile:
```diff
@@ -0,0 +1,59 @@
+  CATEGORY:=Languages
+  TITLE:=LuaSocket 5.3
+  URL:=http://luasocket.luaforge.net/
+  MAINTAINER:=W. Michael Petullo <mike@flyn.org>
```
Is @flyn-org aware of it?

cwluxul replied:
I wasn't sure about this point, since it is an updated package.  I asked one of the other opentwrt developers and they suggested submitting the merge request with the current package maintainer.  I simply wasn't sure how to approach this.

neheb replied:
The original package also needs similar cleanup.

cwluxul replied:
@neheb - how do I go about resolving this (I don't contribute a lot of open source so I don't know all the rules and etiquette).  Do I e-mail the maintainer of the original luasocket package and ask them to maintain this one?  Am I supposed to be maintainer of this new package even though it is a copy of their package?  I apologize for being ignorant on this.

neheb replied:
Whatever you want. Since you're submitting a new package, I welcome you to maintain it. Or leave it empty if you do not want to.

cwluxul replied:
@neheb - I was unaware it was an option to leave maintainer empty.  I've done that.

neheb commented:
hash is wrong. Use make check FIXUP=1

cwluxul commented:
I'm investigating the build failure, my environment is building so I'm trying to understand what is going wrong with the automated build.

karlp commented:
Can we please do something like python2/3 packages are doing, instead of jsut duplicating every single lua package to support both?  This seems like a completely unmaintainable direction.

cwluxul commented:
@karlp 

> Can we please do something like python2/3 packages are doing, instead of jsut duplicating every single lua package to support both? This seems like a completely unmaintainable direction.

@karlp - I agree.  These duplicated packages leave a very bad taste in my mouth.  I also apologize, I don't follow the community enough to know what is going on with the python2/3 packages, nor would I know where to look to come up to speed on what is going on there, but I'll attempt to respond to your comment the best I can.  I think sometime in the future if all the current things that depend on lua 5.1 transition to lua 5.3, the OpenWRT maintainers could deprecate lua 5.1 and ultimately remove it.  I also am NOT an OpenWRT dev expert, but it seems to me that all the things in the OpenWRT system that depend on lua 5.1 make replacing it wholesale a massive task and there are definitely api changes changes between lua 5.1 and lua 5.3.  One example can be seen in this package in the 0001 patch with the replacement of the luaL_openlib function from the luasocket package (5.1) to the luaL_setfuncs function in this luasocket 5.3 package.  

If there is something specific I can do to this package, I'm happy to attempt to accommodate.  If this approach is simply unacceptable and you reject this package I'll figure out a different way to host this package.  This is the only one I know I need for the foreseeable future also, maybe that impacts your decision.

cwluxul commented:
@neheb, @karlp,

I believe I've addressed all the concerns and answered all the questions to the best of my ability, but I'm not sure where this goes from here?  Is there any additional I need to do?

Thanks!

karlp commented:
I'm afraid I don't really see how anything's been addressed. You've replied that "you don't like it" but... that's not really fixing anything. You're still taking this onto a path that copies and pastes the entirety of every lua package.  I don't see how that's remotely maintainable.

rmilecki commented:
I took a moment to look at that packaging issue and I see problem pointed out by @karlp with duplication. It seems that both: `luasocket` and `luasocket5.3` packages can share a lot of code: `PKG_` variables, main `define`, description and patches (as both use the same sources).

So my suggestion is to integrate changes from this Pull Request into an existing `luasocket` package.

@cwluxul: you will need:
* `VARIANT` in both `Package/luasocket` and `Package/luasocket5.3`
* `Build/Compile` modified to use `$BUILD_VARIANT`, I suggest just calling `$(Build/Compile/$(BUILD_VARIANT))` and providing relevant defines
* Work on `0001-Add-interface-support.patch` by either:
  * making it work with both versions
  * using some `#if` or `#ifdef` in it for version specific code
  * extending it by an extra patch applied to `luasocket5.3` only

rmilecki commented:
Obsoleted by #9566
